In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate

import warnings; warnings.simplefilter('ignore')

We can use ast.literal_eval() here to evaluate the string as a python expression. It safely evaluates an expression node or a string containing a Python expression.

Surprise is a Python scikit for building and analyzing recommender systems that deal with explicit rating data.

Top N Recommendations

In [2]:
md = pd.read_csv(r'file:///C:\Users\Ruchi\Downloads\movies_metadata.csv\movies_metadata.csv')
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [3]:
md['genres'] = md['genres'].fillna('[]')
md.head(100)

adult                              belongs_to_collection    budget  \
0   False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1   False                                                NaN  65000000   
2   False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3   False                                                NaN  16000000   
4   False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
..    ...                                                ...       ...   
95  False                                                NaN   3000000   
96  False                                                NaN         0   
97  False                                                NaN         0   
98  False                                                NaN         0   
99  False                                                NaN   7000000   

                                               genres  \
0   [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1   [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2   [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3   [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                      [{'id': 35, 'name': 'Comedy'}]   
..                                                ...   
95                      [{'id': 18, 'name': 'Drama'}]   
96  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
97                [{'id': 99, 'name': 'Documentary'}]   
98  [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...   
99  [{'id': 35, 'name': 'Comedy'}, {'id': 80, 'nam...   

                                          homepage     id    imdb_id  \
0             http://toystory.disney.com/toy-story    862  tt0114709   
1                                              NaN   8844  tt0113497   
2                                              NaN  15602  tt0113228   
3                                              NaN  31357  tt0114885   
4                                              NaN  11862  tt0113041   
..                                             ...    ...        ...   
95                                             NaN    406  tt0113247   
96                                             NaN  45549  tt0111173   
97  http://www.nickbroomfield.com/heidifleiss.html  63076  tt0113283   
98                                             NaN  11062  tt0115907   
99                                             NaN  13685  tt0115734   

   original_language                 original_title  \
0                 en                      Toy Story   
1                 en                        Jumanji   
2                 en               Grumpier Old Men   
3                 en              Waiting to Exhale   
4                 en    Father of the Bride Part II   
..               ...                            ...   
95                fr                       La Haine   
96                en                       Shopping   
97                en  Heidi Fleiss: Hollywood Madam   
98                en                      City Hall   
99                en                  Bottle Rocket   

                                             overview  ... release_date  \
0   Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1   When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2   A family wedding reignites the ancient feud be...  ...   1995-12-22   
3   Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4   Just when George Banks has recovered from his ...  ...   1995-02-10   
..                                                ...  ...          ...   
95  Aimlessly whiling away their days in the concr...  ...   1995-05-31   
96  A dark, hip, urban story of a barren and anony...  ...   1994-12-06   
97  A documentary crew from the BBC arrives in L.A...  ...   1995-12-27   
98  The accidental shooting of a boy in New York l...  ...   1996-02-16   
99  Upon his release from a mental hospital follow...  ...   1996-02-21   

   

Eval example

our genre data is in the form of text, so we cannot extract values by indexing for that we neeed to perform eval function

In [4]:
#eval example

list1 = '[1,2,3,4]'
print(type(list1))
print(list1[0])  

<class 'str'>
[


In [5]:
list_eval = eval(list1)
print(list_eval)

[1, 2, 3, 4]


In [6]:
#eval the data

md['genres'] = md['genres'].apply(literal_eval)

In [7]:
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

#Genres as List 

In [8]:
md['genres'] = md['genres'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

Here we want only genres and not the ids with it so we use the above function to remove that. Now we have converted it into a proper list

In [9]:
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                         genres                              homepage     id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story    862   
1  [Adventure, Fantasy, Family]                                   NaN   8844   
2             [Romance, Comedy]                                   NaN  15602   
3      [Comedy, Drama, Romance]                                   NaN  31357   
4                      [Comedy]                                   NaN  11862   

     imdb_id original_language               original_title  \
0  tt0114709                en                    Toy Story   
1  tt0113497                en                      Jumanji   
2  tt0113228                en             Grumpier Old Men   
3  tt0114885                en            Waiting to Exhale   
4  tt0113041                en  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [10]:
md[md['vote_count'].notnull()] 

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2      A family wedding reignites the ancient feud be...  ...   1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4      Just when George Banks has recovered from his ...  ...   1995-02-10   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisation of homosexual...  ...   2017-06-09   

           revenue runtime                                   spoken_languages  \
0      373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1      262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2              0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}

In [11]:
vote_count = md[md['vote_count'].notnull()]['vote_count'].astype('int') #we want vote counts to be integer values
vote_count 

0        5415
1        2413
2          92
3          34
4         173
         ... 
45461       1
45462       3
45463       6
45464       0
45465       0
Name: vote_count, Length: 45460, dtype: int32

In [12]:
vote_average = md[md['vote_average'].notnull()]['vote_average'].astype('int') #vote average also we want as integer values
vote_average 

0        7
1        6
2        6
3        6
4        5
        ..
45461    4
45462    9
45463    3
45464    0
45465    0
Name: vote_average, Length: 45460, dtype: int32

In [13]:
top_movies=md.copy()

In [14]:
top_movies1 = top_movies.sort_values('vote_average' , ascending=False).head(250) 

i want to give recommendation based on average votes, if votes are more the great movie and votes are less then not a good movie. ascending false because sorting is by default ascending and we dont want that now.

#No Minimum Votes Requirement 

In [15]:
top_movies1

adult belongs_to_collection  budget  \
21642  False                   NaN       0   
15710  False                   NaN       0   
22396  False                   NaN       0   
22395  False                   NaN       0   
35343  False                   NaN  300000   
...      ...                   ...     ...   
35428  False                   NaN       0   
24882  False                   NaN       0   
4183   False                   NaN       0   
36345  False                   NaN       0   
30212  False                   NaN       0   

                                       genres  \
21642                           [Documentary]   
15710                           [Documentary]   
22396                           [Documentary]   
22395                           [Documentary]   
35343  [Comedy, Documentary, Music, TV Movie]   
...                                       ...   
35428                        [Drama, Romance]   
24882                                      []   
4183                        [Thriller, Drama]   
36345                        [History, Drama]   
30212                           [Documentary]   

                                  homepage      id    imdb_id  \
21642                                  NaN  320849  tt0886500   
15710                                  NaN   96451  tt1587373   
22396                                  NaN   72123  tt1341746   
22395  http://www.marvinhamlischmovie.com/  230864  tt3011874   
35343                                  NaN  140595  tt0308213   
...                                    ...     ...        ...   
35428        http://www.kuraitokorode.com/  206155  tt0872007   
24882                                  NaN  273334  tt3575800   
4183                                   NaN   88727  tt0157411   
36345                                  NaN  359154  tt4699592   
30212                                  NaN  300179  tt3402078   

      original_language                                    original_title  \
21642                en   Ice Age Columbus: Who Were the First Americans?   
15710                en         If God Is Willing and da Creek Don't Rise   
22396                en                                    Meat the Truth   
22395                en             Marvin Hamlisch: What He Did For Love   
35343                en                        Elaine Stritch: At Liberty   
...                 ...                                               ...   
35428                ja                                       暗いところで待ち合わせ   
24882                en  United States of Secrets (Part One): The Program   
4183                 no                                       1732 Høtten   
36345                ab                                13 dies de octubre   
30212                en                            The Lion’s Mouth Opens   

                                                overview  ... release_date  \
21642  Firmly rooted in the latest scientific discove...  ...   2005-01-01   
15710  In 2006, director Spike Lee created an astonis...  ...   2010-08-23   
22396  Meat the Truth is a high-profile documentary w...  ...   2008-10-03   
22395  When Marvin Hamlisch passed away in August 201...  ...   2013-10-12   
35343  Judy at the Palace. Sinatra at Carnegie Hall. ...  ...   2002-01-01   
...                                                  ...  ...          ...   
35428  Michiru has lost her eyesight, and to make mat...  ...   2006-09-05   
24882  Last year Edward Snowden downloaded tens of th...  ...   2014-05-13   
4183   A few months after a girl with developmental d...  ...   1998-12-26   
36345                                                NaN  ...   2015-09-10   
30212  A documentary on filmmaker Marianna Palka, as ...  ...   2014-01-18   

      revenue runtime                                   spoken_languages  \
21642     0.0     0.0                                                 []   
15710     0.0   255.0                                                 

In [16]:
##Minimum Number of Votes 1000

In [17]:
top_movies2 = top_movies[top_movies['vote_count']>1000]

In [18]:
top_movies2

adult                              belongs_to_collection     budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...   30000000   
1      False                                                NaN   65000000   
5      False                                                NaN   60000000   
9      False  {'id': 645, 'name': 'James Bond Collection', '...   58000000   
15     False                                                NaN   52000000   
...      ...                                                ...        ...   
43644  False                                                NaN   34000000   
44009  False  {'id': 86066, 'name': 'Despicable Me Collectio...   80000000   
44274  False  {'id': 173710, 'name': 'Planet of the Apes (Re...  152000000   
44678  False                                                NaN  100000000   
44842  False  {'id': 8650, 'name': 'Transformers Collection'...  260000000   

                                               genres  \
0                         [Animation, Comedy, Family]   
1                        [Adventure, Fantasy, Family]   
5                    [Action, Crime, Drama, Thriller]   
9                       [Adventure, Action, Thriller]   
15                                     [Drama, Crime]   
...                                               ...   
43644                                 [Action, Crime]   
44009  [Action, Animation, Adventure, Family, Comedy]   
44274                   [Drama, Science Fiction, War]   
44678         [Action, Drama, History, Thriller, War]   
44842  [Action, Science Fiction, Thriller, Adventure]   

                                                homepage      id    imdb_id  \
0                   http://toystory.disney.com/toy-story     862  tt0114709   
1                                                    NaN    8844  tt0113497   
5                                                    NaN     949  tt0113277   
9           http://www.mgm.com/view/movie/757/Goldeneye/     710  tt0113189   
15                                                   NaN     524  tt0112641   
...                                                  ...     ...        ...   
43644                                                NaN  339403  tt3890160   
44009                           http://www.despicable.me  324852  tt3469046   
44274  http://www.foxmovies.com/movies/war-for-the-pl...  281338  tt3450958   
44678                       http://www.dunkirkmovie.com/  374720  tt5013056   
44842                  http://www.transformersmovie.com/  335988  tt3371366   

      original_language                  original_title  \
0                    en                       Toy Story   
1                    en                         Jumanji   
5                    en                            Heat   
9                    en                       GoldenEye   
15                   en                          Casino   
...                 ...                             ...   
43644                en                     Baby Driver   
44009                en                 Despicable Me 3   
44274                en  War for the Planet of the Apes   
44678                en                         Dunkirk   
44842                en   Transformers: The Last Knight   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
5      Obsessive master thief, Neil McCauley leads a ...  ...   1995-12-15   
9      James Bond must unmask the mysterious head of ...  ...   1995-11-16   
15     The life of the gambling paradise – Las Vegas ...  ...   1995-11-22   
...                                                  ...  ...          ...   
43644  After being coerced into working for a crime b...  ...   2017-06-28   
44009  Gru and his wife Lucy must stop former '80s ch...  ...   2017-06-15   
44274  Caesar and his ape

weighted rating (W) = R.v + C.m/v+m

where: R = average for the movie as number from 1 to 10 (mean) = (Rating) v = number of votes for the movie = (votes) m = minimum votes required to be listed in the Top 250 (currently 25,000) C = the mean vote across the whole report (currently 7.0)

The next step is to determine an appropriate value for m, the minimum votes required to be listed in the chart. We will use 95th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list. We will Calculate C value as average vote across the dataset. Here we Build our overall Top 250 Chart and will define a function to build charts for a particular genre.

In [19]:
#for getting the mean value of vote count
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_average.mean()
C

5.244896612406511

In [20]:
m = vote_counts.quantile(0.95)  
m

434.0

In [21]:
top_movies['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [22]:
top_movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ...      revenue  \
0      Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1      When siblings Judy and Peter discover an encha...  ...  262797249.0   
2      A family wedding reignites the ancient feud be...  ...          0.0   
3      Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4      Just when George Banks has recovered from his ...  ...   76578911.0   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          0.0   
45462  An artist struggles to finish his work while a...  ...          0.0   
45463  When one of her hits goes wrong, a professiona...  ...          0.0   
45464  In a small town live two brothers, one a minis...  ...          0.0   
45465  50 years after decriminalisation of homosexual...  ...          0.0   

      runtime                                   spoken_languages    status  \
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1       104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2       101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3      

In [23]:
top_movies3 = top_movies[(top_movies['vote_count'] >= m) & (top_movies['vote_count'].notnull()) & (top_movies['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
top_movies3['vote_count'] = top_movies3['vote_count'].astype('int')
top_movies3['vote_average'] = top_movies3['vote_average'].astype('int')
top_movies3.shape

(2274, 6)

In [24]:
 #for getting top movies
    
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [25]:
top_movies3['weight_rate'] = top_movies3.apply(weighted_rating, axis=1)

In [26]:
top_movies3.head()

title  year  vote_count  vote_average popularity  \
0   Toy Story  1995        5415             7    21.9469   
1     Jumanji  1995        2413             6    17.0155   
5        Heat  1995        1886             7    17.9249   
9   GoldenEye  1995        1194             6     14.686   
15     Casino  1995        1343             7    10.1374   

                              genres  weight_rate  
0        [Animation, Comedy, Family]     6.869770  
1       [Adventure, Fantasy, Family]     5.884891  
5   [Action, Crime, Drama, Thriller]     6.671675  
9      [Adventure, Action, Thriller]     5.798701  
15                    [Drama, Crime]     6.571348

In [27]:
top_movies3 = top_movies3.sort_values('weight_rate', ascending=False).head(10)

In [28]:
top_movies3.head(10)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22879                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
292                                         Pulp Fiction  1994        8670   
314                             The Shawshank Redemption  1994        8358   
7000       The Lord of the Rings: The Return of the King  2003        8226   
351                                         Forrest Gump  1994        8147   
5814               The Lord of the Rings: The Two Towers  2002        7641   

       vote_average popularity  \
15480             8    29.1081   
12481             8    123.167   
22879             8    32.2135   
2843              8    63.8696   
4863              8    32.0707   
292               8     140.95   
314               8    51.6454   
7000              8    29.3244   
351               8    48.3072   
5814              8    29.4235   

                                                  genres  weight_rate  
15480  [Action, Thriller, Science Fiction, Mystery, A...     7.917588  
12481                   [Drama, Action, Crime, Thriller]     7.905871  
22879                [Adventure, Drama, Science Fiction]     7.897107  
2843                                             [Drama]     7.881753  
4863                        [Adventure, Fantasy, Action]     7.871787  
292                                    [Thriller, Crime]     7.868660  
314                                       [Drama, Crime]     7.864000  
7000                        [Adventure, Fantasy, Action]     7.861927  
351                             [Comedy, Drama, Romance]     7.860656  
5814                        [Adventure, Fantasy, Action]     7.851924

# Top movies

In [29]:
## Genre = Romance

In [30]:
#pre processed the data to sepearate all the genres and not in one list

genre_TM = top_movies.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
genre_TM.name = 'genre'
genre_top_movies= top_movies.drop('genres', axis=1).join(genre_TM)

In [31]:
genre_top_movies.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
1  False                                                NaN  65000000   

                               homepage    id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
0  http://toystory.disney.com/toy-story   862  tt0114709                en   
1                                   NaN  8844  tt0113497                en   
1                                   NaN  8844  tt0113497                en   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   
1        Jumanji  When siblings Judy and Peter discover an encha...   
1        Jumanji  When siblings Judy and Peter discover an encha...   

  popularity  ... runtime                                   spoken_languages  \
0    21.9469  ...    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
0    21.9469  ...    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
0    21.9469  ...    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1    17.0155  ...   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
1    17.0155  ...   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   

     status                                    tagline      title  video  \
0  Released                                        NaN  Toy Story  False   
0  Released                                        NaN  Toy Story  False   
0  Released                                        NaN  Toy Story  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   
1  Released  Roll the dice and unleash the excitement!    Jumanji  False   

  vote_average vote_count  year      genre  
0          7.7     5415.0  1995  Animation  
0          7.7     5415.0  1995     Comedy  
0          7.7     5415.0  1995     Family  
1          6.9     2413.0  1995  Adventure  
1          6.9     2413.0  1995    Fantasy  

[5 rows x 25 columns]

In [32]:
 #it will paste the result in specific genre respectively
    
def build_chart(genre, percentile=0.85):
    df = genre_top_movies[genre_top_movies['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

In [33]:
build_chart('Comedy').head(15)

title  year  vote_count  \
10309                        Dilwale Dulhania Le Jayenge  1995         661   
351                                         Forrest Gump  1994        8147   
1225                                  Back to the Future  1985        6239   
18465                                   The Intouchables  2011        5410   
22841                           The Grand Budapest Hotel  2014        4644   
2211                                   Life Is Beautiful  1997        3643   
732    Dr. Strangelove or: How I Learned to Stop Worr...  1964        1472   
3342                                        Modern Times  1936         881   
883                                     Some Like It Hot  1959         835   
1236                                  The Great Dictator  1940         756   
882                                        The Apartment  1960         498   
3189                                         City Lights  1931         444   
31658                                              Feast  2014         420   
3192                                             The Kid  1921         404   
2906                                         The General  1926         240   

       vote_average popularity        wr  
10309             9     34.457  8.463024  
351               8    48.3072  7.963363  
1225              8    25.7785  7.952358  
18465             8    16.0869  7.945207  
22841             8     14.442  7.936384  
2211              8     39.395  7.919430  
732               8    9.80398  7.809073  
3342              8    8.15956  7.695554  
883               8    11.8451  7.680781  
1236              8    9.24175  7.651762  
882               8    11.9943  7.504715  
3189              8    10.8915  7.456698  
31658             8    7.36566  7.432234  
3192              8    8.16846  7.414663  
2906              8    8.00295  7.142722

In [34]:
build_chart('Family').head(15)

title  year  vote_count  \
1225                         Back to the Future  1985        6239   
359                               The Lion King  1994        5520   
5481                              Spirited Away  2001        3968   
5833                         My Neighbor Totoro  1988        1730   
926                       It's a Wonderful Life  1946        1103   
19901                                  Paperman  2012         734   
39386                                     Piper  2016         487   
20779                             Wolf Children  2012         483   
31658                                     Feast  2014         420   
25044                           Song of the Sea  2014         420   
4766   Harry Potter and the Philosopher's Stone  2001        7188   
13724                                        Up  2009        7048   
30315                                Inside Out  2015        6737   
15472                             Despicable Me  2010        6595   
12704                                    WALL·E  2008        6439   

       vote_average popularity        wr  
1225              8    25.7785  7.893053  
359               8    21.6058  7.879754  
5481              8    41.0489  7.835635  
5833              8    13.5073  7.650968  
926               8    15.0316  7.490637  
19901             8    7.19863  7.301918  
39386             8  11.243161  7.071694  
20779             8    10.2495  7.066709  
31658             8    7.36566  6.980490  
25044             8    6.96736  6.980490  
4766              7    38.1872  6.941838  
13724             7    19.3309  6.940724  
30315             7    23.9856  6.938090  
15472             7    22.2745  6.936808  
12704             7    16.0884  6.935337

In [35]:
build_chart('Action').head(15)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
7000       The Lord of the Rings: The Return of the King  2003        8226   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1154                             The Empire Strikes Back  1980        5998   
4135                                            Scarface  1983        3017   
9430                                              Oldboy  2003        2000   
1910                                       Seven Samurai  1954         892   
43190                                   Band of Brothers  2001         725   
1215                                                   M  1931         465   
14551                                             Avatar  2009       12114   
17818                                       The Avengers  2012       12000   
26564                                           Deadpool  2016       11444   

       vote_average popularity        wr  
15480             8    29.1081  7.955099  
12481             8    123.167  7.948610  
4863              8    32.0707  7.929579  
7000              8    29.3244  7.924031  
5814              8    29.4235  7.918382  
256               8    42.1497  7.908327  
1154              8     19.471  7.896841  
4135              8    11.2997  7.802046  
9430              8    10.6169  7.711649  
1910              8    15.0178  7.426145  
43190             8   7.903731  7.325485  
1215              8    12.7524  7.072073  
14551             7    185.071  6.966363  
17818             7    89.8876  6.966049  
26564             7     187.86  6.964431

 Content Based Recommender

In [36]:
links_small = pd.read_csv(r'file:///C:\Users\Ruchi\Downloads\links_small.csv')
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')

In [37]:
top_movies = top_movies.drop([19730, 29503, 35587])

In [38]:
top_movies['id'] = top_movies['id'].astype('int')  #Check EDA Notebook for how and why we got these indices.
top_movies

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                             genres                              homepage  \
0       [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   
1      [Adventure, Fantasy, Family]                                   NaN   
2                 [Romance, Comedy]                                   NaN   
3          [Comedy, Drama, Romance]                                   NaN   
4                          [Comedy]                                   NaN   
...                             ...                                   ...   
45461               [Drama, Family]  http://www.imdb.com/title/tt6209470/   
45462                       [Drama]                                   NaN   
45463     [Action, Drama, Thriller]                                   NaN   
45464                            []                                   NaN   
45465                            []                                   NaN   

           id    imdb_id original_language               original_title  \
0         862  tt0114709                en                    Toy Story   
1        8844  tt0113497                en                      Jumanji   
2       15602  tt0113228                en             Grumpier Old Men   
3       31357  tt0114885                en            Waiting to Exhale   
4       11862  tt0113041                en  Father of the Bride Part II   
...       ...        ...               ...                          ...   
45461  439050  tt6209470                fa                      رگ خواب   
45462  111109  tt2028550                tl          Siglo ng Pagluluwal   
45463   67758  tt0303758                en                     Betrayal   
45464  227506  tt0008536                en          Satana likuyushchiy   
45465  461257  tt6980792                en                     Queerama   

                                                overview  ...      revenue  \
0      Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1      When siblings Judy and Peter discover an encha...  ...  262797249.0   
2      A family wedding reignites the ancient feud be...  ...          0.0   
3      Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4      Just when George Banks has recovered from his ...  ...   76578911.0   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          0.0   
45462  An artist struggles to finish his work while a...  ...          0.0   
45463  When one of her hits goes wrong, a professiona...  ...          0.0   
45464  In a small town live two brothers, one a minis...  ...          0.0   
45465  50 years after decriminalisation of homosexual...  ...          0.0   

      runtime                                   spoken_languages    status  \
0        81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1       104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2       101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3      

In [39]:
top_movies4 = top_movies[top_movies['id'].isin(links_small)]
top_movies4.shape

(9099, 25)

In [40]:
top_movies4.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                         genres                              homepage     id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story    862   
1  [Adventure, Fantasy, Family]                                   NaN   8844   
2             [Romance, Comedy]                                   NaN  15602   
3      [Comedy, Drama, Romance]                                   NaN  31357   
4                      [Comedy]                                   NaN  11862   

     imdb_id original_language               original_title  \
0  tt0114709                en                    Toy Story   
1  tt0113497                en                      Jumanji   
2  tt0113228                en             Grumpier Old Men   
3  tt0114885                en            Waiting to Exhale   
4  tt0113041                en  Father of the Bride Part II   

                                            overview  ...      revenue  \
0  Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1  When siblings Judy and Peter discover an encha...  ...  262797249.0   
2  A family wedding reignites the ancient feud be...  ...          0.0   
3  Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4  Just when George Banks has recovered from his ...  ...   76578911.0   

  runtime                                   spoken_languages    status  \
0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2   101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3   127.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
4   106.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  year  
0                    Toy Story  False          7.7     5415.0  1995  
1                      Jumanji  False          6.9     2413.0  1995  
2             Grumpier Old Men  False          6.5       92.0  1995  
3            Waiting to Exhale  False          6.1       34.0  1995  
4  Father of the Bride Part II  False          5.7      173.0  1995  

[5 rows x 25 columns]

We have 9099 movies available in our small movies metadata dataset which is 5 times smaller than our original data set of 450000 movies.

Movie Description Based Recommender

In [41]:
top_movies4['tagline'] = top_movies4['tagline'].fillna('')
top_movies4['description'] = top_movies4['overview'] + top_movies4['tagline']
top_movies4['description'] = top_movies4['description'].fillna('')

In [42]:
 tf = TfidfVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
tfidf_matrix = tf.fit_transform(top_movies4['description'])

In [43]:
tfidf_matrix.shape

(9099, 268124)

Cosine Similarity: I will be using the Cosine Similarity to calculate a numeric quantity that denotes the similarity between two movies. Mathematically, it is defined as follows:

cosine(x,y)=(x.y⊺) ÷ ||x||.||y|| Since we have used the TF-IDF Vectorizer, calculating the Dot Product will directly give us the Cosine Similarity Score. Therefore, we will use sklearn's linear_kernel instead of cosine_similarities since it is much faster.

In [44]:
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [45]:
cosine_sim

array([[1.        , 0.00680476, 0.        , ..., 0.        , 0.00344913,
        0.        ],
       [0.00680476, 1.        , 0.01531062, ..., 0.00357057, 0.00762326,
        0.        ],
       [0.        , 0.01531062, 1.        , ..., 0.        , 0.00286535,
        0.00472155],
       ...,
       [0.        , 0.00357057, 0.        , ..., 1.        , 0.07811616,
        0.        ],
       [0.00344913, 0.00762326, 0.00286535, ..., 0.07811616, 1.        ,
        0.        ],
       [0.        , 0.        , 0.00472155, ..., 0.        , 0.        ,
        1.        ]])

In [46]:
cosine_sim[0] 

array([1.        , 0.00680476, 0.        , ..., 0.        , 0.00344913,
       0.        ])

We now have a pairwise cosine similarity matrix for all the movies in our dataset.

In [47]:
top_movies4 = top_movies4.reset_index()
titles = top_movies4['title']
indices = pd.Series(top_movies4.index, index=top_movies4['title'])

In [48]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

In [49]:
get_recommendations('GoldenEye').head(10)   #now try and get the top recommendations for a few movies and see how good the recommendations are.

5172                          Octopussy
2398                   Live and Let Die
4309                      Casino Royale
5175              Never Say Never Again
2396                 For Your Eyes Only
1883                   A View to a Kill
2901        The Man with the Golden Gun
5171                You Only Live Twice
2397                    Licence to Kill
2896    On Her Majesty's Secret Service
Name: title, dtype: object

In [50]:
get_recommendations('The Apartment').head(10)

1925                          Nothing in Common
5242                     Shadow of the Thin Man
5769                          The Holy Mountain
3904                                    48 Hrs.
3221                         The House of Mirth
5358                               Safety Last!
1590                     The Barefoot Executive
1961                             Running Scared
1003              The Day the Earth Stood Still
75      Things to Do in Denver When You're Dead
Name: title, dtype: object

In [51]:
get_recommendations('The Godfather').head(10)

973      The Godfather: Part II
8387                 The Family
3509                       Made
4196         Johnny Dangerously
29               Shanghai Triad
5667                       Fury
2412             American Movie
1582    The Godfather: Part III
4221                    8 Women
2159              Summer of Sam
Name: title, dtype: object

In [52]:
get_recommendations('The Dark Knight').head(10)

7931                      The Dark Knight Rises
132                              Batman Forever
1113                             Batman Returns
8227    Batman: The Dark Knight Returns, Part 2
7565                 Batman: Under the Red Hood
524                                      Batman
7901                           Batman: Year One
2579               Batman: Mask of the Phantasm
2696                                        JFK
8165    Batman: The Dark Knight Returns, Part 1
Name: title, dtype: object